controllers/integration/integration_adapter.go

```diff
@@ -0,0 +1,310 @@
+		context:     context,
+	}
+}
+
+// EnsureApplicationSnapshotExists is an operation that will ensure that an integration ApplicationSnapshot associated
+// to the PipelineRun being processed exists. Otherwise, it will create a new integration ApplicationSnapshot.
+func (a *Adapter) EnsureApplicationSnapshotExists() (results.OperationResult, error) {
+	existingApplicationSnapshot, err := a.findMatchingApplicationSnapshot()
+	if err != nil {
+		return results.RequeueWithError(err)
+	}
+
+	if existingApplicationSnapshot == nil {
+		newApplicationSnapshot, err := a.createApplicationSnapshotForPipelineRun()
+		if err != nil {
+			a.logger.Error(err, "Failed to create ApplicationSnapshot",
+				"Application.Name", a.application.Name, "Application.Namespace", a.application.Namespace)
+			return results.RequeueOnErrorOrStop(a.updateStatus())
+		}
+		err = a.client.Create(a.context, newApplicationSnapshot)
+		if err != nil {
+			a.logger.Error(err, "Failed to create ApplicationSnapshot",
+				"Application.Name", a.application.Name, "Application.Namespace", a.application.Namespace)
+			return results.RequeueOnErrorOrStop(a.updateStatus())
+		}
+		a.logger.Info("Created new ApplicationSnapshot",
+			"Application.Name", a.application.Name,
+			"ApplicationSnapshot.Name", newApplicationSnapshot.Name,
+			"ApplicationSnapshot.Spec.Images", newApplicationSnapshot.Spec.Images)
+
+	} else {
+		a.logger.Info("Found existing ApplicationSnapshot",
+			"Application.Name", a.application.Name,
+			"ApplicationSnapshot.Name", existingApplicationSnapshot.Name,
+			"ApplicationSnapshot.Spec.Images", existingApplicationSnapshot.Spec.Images)
+	}
+
+	return results.ContinueProcessing()
+}
+
+// EnsureAllReleasesExist is an operation that will ensure that all integration Releases associated
+// to the ApplicationSnapshot and the Application's ReleaseLinks exist.
+// Otherwise, it will create new Releases for each ReleaseLink.
+func (a *Adapter) EnsureAllReleasesExist() (results.OperationResult, error) {
+	existingApplicationSnapshot, err := a.findMatchingApplicationSnapshot()
+	if err != nil {
+		return results.RequeueWithError(err)
+	} else if existingApplicationSnapshot == nil {
+		return results.RequeueAfter(time.Second*5, err)
+	}
+
+	releaseLinks, err := a.getAllApplicationReleaseLinks()
+	if err != nil {
+		a.logger.Error(err, "Failed to get all ReleaseLinks",
+			"Application.Name", a.application.Name,
+			"Application.Namespace", a.application.Namespace)
+		return results.RequeueOnErrorOrStop(a.updateStatus())
+	}
+
+	releases, err := a.getAllApplicationSnapshotReleases(existingApplicationSnapshot)
+	if err != nil {
+		a.logger.Error(err, "Failed to get all Releases",
+			"ApplicationSnapshot.Name", existingApplicationSnapshot.Name,
+			"ApplicationSnapshot.Namespace", existingApplicationSnapshot.Namespace)
+		return results.RequeueOnErrorOrStop(a.updateStatus())
+	}
+	for _, releaseLink := range *releaseLinks {
+		var existingRelease *releasev1alpha1.Release = nil
```

Comment:
I'd extract the whole loop content into a function for readability.

Reply:
Ack, added the function to handle this